### PR TITLE
Implement rewrite of inventory queries.

### PR DESCRIPTION
**Pull request by JBYoshi**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1610) | **SpongeCommon**

When inventories were first added, they had support for querying for sub-inventories based on various attributes about them. One of those attributes was the type of `ItemStack` they contained. However, sometimes, developers want to use the quantity of the `ItemStack`, and sometimes they want to ignore it but keep the data on it. Originally, the way to specify this was through a quantity of `-1`. Then Minecraft dropped support for negative quantities. Sponge left in its wake a `queryAny()` method that ignored the quantity. In a discussion earlier today, Mumfrey decided this system needed to change, and I decided to be the one to write it. This version exposes a `QueryOperation` that tells the query how to treat each argument. It should also be more flexible than the current system in case Minecraft breaks some other behavior in the future.

Closes https://github.com/SpongePowered/SpongeCommon/issues/1431.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/item/inventory/query/SpongeQueryOperationType.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    public SpongeQueryOperationType(String id, Function<T, SpongeQueryOperation> newInstance) {
+        super(id);
+        this.newInstance = newInstance;
+    }
+
+    @Override
+    public QueryOperation of(T arg) {
+        checkNotNull(arg);
+        return newInstance.apply(arg);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/query/SpongeQueryOperationType.java]: `this.`

File: src/main/java/org/spongepowered/common/registry/CommonModuleRegistry.java [hunk 2]
```diff
@@ -41,8 +41,51 @@
 import org.spongepowered.api.data.meta.PatternLayer;
 import org.spongepowered.api.data.persistence.DataFormat;
 import org.spongepowered.api.data.persistence.DataTranslator;
-import org.spongepowered.api.data.type.*;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/registry/CommonModuleRegistry.java]: undo

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/registry/CommonModuleRegistry.java (hunk 2 above)
Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/registry/CommonModuleRegistry.java]: I went back through all the imports only to find that IntelliJ undid all of my changes... then I pressed the "undo" button and it immediately went back through and "fixed" the imports for me... :rage:

File: src/main/java/org/spongepowered/common/item/inventory/query/SpongeQueryOperationType.java (hunk 1 above)
Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/item/inventory/query/SpongeQueryOperationType.java]: Fixed.

**Comment by Zidane** [comment]
@JBYoshi Is this ready to go?

@Faithcaio @Mumfrey How is this looking?

**Pull request merged**